### PR TITLE
Properly restart GZDoom in Windows

### PR DESCRIPTION
**Pull request by Cacodemon345**
Wait up to 5 seconds for the previous GZDoom process to close to avoid overwriting the configuration file.

https://forum.zdoom.org/viewtopic.php?f=2&t=69693

### CODE REVIEW

File: src/common/platform/win32/i_main.cpp [hunk 1]
```diff
@@ -560,10 +561,12 @@ INT_PTR CALLBACK ErrorPaneProc (HWND hDlg, UINT msg, WPARAM wParam, LPARAM lPara
 			if (LOWORD(wParam) == IDC_BUTTON1) // we pressed the restart button, so run GZDoom again
 			{
 				HMODULE hModule = GetModuleHandleW(NULL);
+				LPWSTR cmdline = (LPWSTR)calloc(wcslen(GetCommandLineW()) + sizeof("-waitforprocess") + 12, sizeof(WCHAR));
```
Comment (alexey-lysiuk) [inline src/common/platform/win32/i_main.cpp:564]: Memory allocated by `calloc()` must be freed explicitly. That’s why usage of `TArray<>` is preferred.

File: src/common/platform/win32/i_main.cpp [hunk 2]
```diff
@@ -560,10 +561,12 @@ INT_PTR CALLBACK ErrorPaneProc (HWND hDlg, UINT msg, WPARAM wParam, LPARAM lPara
...
 				WCHAR path[MAX_PATH];
-				GetModuleFileNameW(hModule, path, MAX_PATH);
 
-				ShellExecuteW(NULL, L"open", path, GetCommandLineW(), NULL, SW_SHOWNORMAL);
+				swprintf_s(cmdline, wcslen(GetCommandLineW()) + sizeof("-waitforprocess") + 12, L"%s %s %d", GetCommandLineW(), L"-waitforprocess", GetCurrentProcessId());;
```
Comment (alexey-lysiuk) [inline src/common/platform/win32/i_main.cpp:567]: It’s better to calculate the desired buffer size once, and without a magic number, or at least with a comment why it has such value.

File: src/common/platform/win32/i_main.cpp (hunk 1 above)
Comment (coelckers) [reply src/common/platform/win32/i_main.cpp:564]: std::wstring would be preferred here. Please no C allocations unless there's no way around them, I'd rather migrate the entire code base off them.

**Comment by Cacodemon345** [comment]
Those should fix the "Restart" button completely.

**Comment by drfrag666** [comment]
I don't know if 5 seconds it's too much, probably it's okay. Any idea of where and why does the "overwrite" happen? I'd like to go to the root of the problem but i can't find it.

**Comment by madame-rachelle** [comment]
I think a better solution would be file mutexes. I don't really like adding timers, or hacking around the command line with process id's. It is most definitely available in Windows, and may possibly be available for POSIX systems as well, for the eventuality when the start screen gets revamped. If Graf and mental are okay with the solution proposed here, I am not going to raise too much fuss, at least it addresses the problem, but I think getting down to the bare wire with the problem will overall be more robust and reliable for a solution.

**Pull request closed without merging**